### PR TITLE
STACK-1761  Octavia Flavor Schema Regex Name Expressions

### PR DESCRIPTION
**Pull request by ytsai-a10**
## Description
If Feature Addition:
- Required: STACK-1761 Octavia Flavor Schema Regex Name Expressions

The feature description:
https://teams.microsoft.com/l/file/F748C58F-B88D-4647-A888-5BBA28E34176?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=pptx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FGeneral%2FPM%20Updates%2FBlizzard_Soln.pptx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:44c2ee5a73ec44a1a6962a86c1c06442@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

Please also check PR for acos-client part: https://github.com/a10networks/acos-client/pull/314

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1761

## Technical Approach
Given a name expression, when it has been added to a flavor profile, then the name expression will be available to the loadbalancer 
Given a flavor profile, when flavor data is provided that doesn’t match the schema, then an error should be thrown
Given a slb object, when a name expression matches its name, then the related configuration should be applied
Given a flavor and a configuration value, when both are defined, then the flavor will take priority
Given a name-expression and flavor, when both are defined, then the flavor will take priority
Given a name-expression and config, when both are defined, then the name-expression will take priority
Given a loadbalancer with a flavorprofile, when the flavorprofile has been updated with new information, then that information should be passed to the loadbalancer on the next update

## Config Changes
No configuration change needed, since now we using flavor instead.

The flavor used for testing:
 - slb virtual server
<pre>
<b>{ 

    "virtual-server": 

    { 

        "vport-disable-action":"drop-packet" 

    } 

} </b>
</pre>

```
openstack loadbalancer flavorprofile create --name fp1 --provider a10 --flavor-data '{"virtual-server": {"vport-disable-action":"drop-packet"}}' 
openstack loadbalancer flavor create --name f1 --flavorprofile fp1 --description "vtest" --enable 
openstack loadbalancer create --flavor f1 --vip-subnet-id f25ce642-f953-4058-8c46-98fdd72fb129 --vip-address 192.168.91.56 --name vip1 
```

 - slb virtual port, service-group and server
<pre>
<b>{ 

    "virtual-port": { 

        "user-tag": "1", 

        “name-expressions": [ 

            { 

                "regex": "vport1", 

                "json": {"support-http2": 1} 

            }, 

        ] 

    }, 

    "server": { 

        "conn-limit": 65535, 

        “name-expressions": [ 

            { 

                "regex": "srv1", 

                "json": {"conn-resume": 5000} 

            }, 

        ] 

    }, 

    "service-group": { 

        "lb-method": "fastest-response", 

        "name-expressions": [ 

            { 

                "regex": "sg1",  

                "json": {"health-check-disable":1} 

            } 

        ] 

    } 

}</b>
</pre>

```
openstack loadbalancer flavorprofile create --name fp_all1 --provider a10 --flavor-data '{"virtual-port": {"user-tag": "1", "name-expressions": [{"regex": "vport1", "json": {"support-http2": 1}},]},"server": {"conn-limit": 65535, "name-expressions": [{"regex": "srv1",    "json": {"conn-resume": 5000}},]}, "service-group": {"lc-method": "least-connection", "name-expressions": [{"regex": "sg1", "json": {"health-check-disable":1}}]}}' 
openstack loadbalancer flavor create --name f_all1 --flavorprofile fp_all1 --description "flaovr all test1" --enable 
openstack loadbalancer create --flavor f_all1 --vip-subnet-id f25ce642-f953-4058-8c46-98fdd72fb129 --vip-address 192.168.91.56 --name vip1 
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1 
openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener vport1 --name sg1 
openstack loadbalancer member create --address 192.168.90.132 --subnet-id tp90 --protocol-port 80 --name srv1 sg1 
openstack loadbalancer member create --address 192.168.90.132 --subnet-id tp90 --protocol-port 8080 --weight 2 --name srv1 sg1 
```

- health monitor
<pre>
<b>{ 

  "health-monitor":{ 

    "monitor":{ 

      "retry":5, 

      "method":{ 

        "http":{ 

          "http-response-code":"201" 

        } 

      } 

    }, 

    "name-expressions":[ 

    { 

      "regex":"hm1", 

      "json":{ 

        "monitor":{ 

          "timeout":8, 

          "method":{ 

            "http":{ 

              "http-host":"my.test.com" 

            } 

          } 

        } 

      } 

    } 

    ] 

  } 

} </b>
</pre>

```
openstack loadbalancer flavorprofile create --name fp_hm1 --provider a10 --flavor-data '{"health-monitor": {"monitor": {"retry":5, "method": { "http": {"http-response-code":"201"}}}, "name-expressions": [{"regex": "hm1", "json": {"monitor": {"timeout":8, "method": { "http": {"http-host":"my.test.com"}}}}}]}}' 
openstack loadbalancer flavor create --name f_hm1 --flavorprofile fp_hm1 --description "hm test1" --enable 
openstack loadbalancer create --flavor f_hm1 --vip-subnet-id f25ce642-f953-4058-8c46-98fdd72fb129 --vip-address 192.168.91.56 --name vip1 
openstack loadbalancer listener create --protocol HTTP --protocol-port 80 --name vport1 vip1 
openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener vport1 --name sg1 
openstack loadbalancer healthmonitor create --delay 30 --timeout 3 --max-retries 3 --type HTTP sg1 --name hm1 
```
## Test Cases
 - flavor support for loadbalancer create
 - flavor support for listener create
 - flavor support for pool create
 - flavor support for member create
 - flavor support for healthmonitor create
 - flavor support for loadbalancer set
 - flavor support for listener set
 - flavor support for pool set
 - flavor support for member set
 - flavor support for healthmonitor set

## Manual Testing
Please find the unit-test document in this link, it will provide the test description, flavor information, openstack commands for test and result:
https://teams.microsoft.com/l/file/E84B76C6-48D8-441C-B375-51B010F42EB3?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2FFlavor%20Support%2FRD%20unit-test%20notes.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

### CODE REVIEW

File: a10_octavia/api/drivers/driver.py [hunk 1]
```diff
@@ -12,12 +12,19 @@
...
 #    under the License.
 
+from jsonschema import exceptions as js_exceptions
+from jsonschema import validate
 from oslo_config import cfg
 from oslo_log import log as logging
 import oslo_messaging as messaging
 
-from octavia.api.drivers import provider_base as driver_base
+# from octavia.api.drivers import provider_base as driver_base
```
Comment (afrin-chakure-a10) [inline a10_octavia/api/drivers/driver.py:21]: You can remove

File: ChangeLog [hunk 2]
```diff
@@ -1,6 +1,598 @@
 CHANGES
 =======
 
+* Fixed positive testcase for GetFlavorObject execute method
```
Comment (afrin-chakure-a10) [inline ChangeLog:4]: You can remove this file

File: AUTHORS [hunk 3]
```diff
@@ -5,11 +5,25 @@ Michael Durrant <mdurrant-b3@users.noreply.github.com>
 Michael Durrant <mdurrant@a10networks.com>
```
Comment (afrin-chakure-a10) [inline AUTHORS:5]: You can remove this file

File: a10_octavia/db/migration/alembic_migrations/versions/05b1446c7f20_add_subnet_id_column_to_vrid_table.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 # revision identifiers, used by Alembic.
 revision = '05b1446c7f20'
-down_revision = '4028e5f7a198'
+down_revision = 'ea8bfcbd654c'
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/05b1446c7f20_add_subnet_id_column_to_vrid_table.py:14]: You can remove this file as it not related to Flavor feature

File: a10_octavia/db/migration/alembic_migrations/versions/b02c78488551_updating_vthunders_table_with_new_column.py [hunk 5]
```diff
@@ -17,7 +17,7 @@
 
 
 def upgrade():
-    op.add_column('vthunders', sa.Column('hierarchical_multitenancy', sa.String(7), default=None, nullable=True))
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/b02c78488551_updating_vthunders_table_with_new_column.py:20]: You can remove this file as it not related to Flavor feature

File: a10_octavia/db/migration/alembic_migrations/versions/ea8bfcbd654c_modified_hmt_column_to_accept_only_.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+"""Modified hmt column to accept only enable/disable
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/migration/alembic_migrations/versions/ea8bfcbd654c_modified_hmt_column_to_accept_only_.py:1]: You can remove this file as it not related to Flavor feature

File: a10_octavia/db/models.py [hunk 7]
```diff
@@ -18,6 +18,7 @@
 from sqlalchemy import orm
 from sqlalchemy.orm import validates
 from sqlalchemy.sql import func
+from sqlalchemy_utils.types.choice import ChoiceType
```
Comment (afrin-chakure-a10) [inline a10_octavia/db/models.py:21]: You can remove this file as it not related to Flavor feature

File: a10_octavia/tests/common/a10constants.py [hunk 8]
```diff
@@ -51,6 +53,7 @@
 LISTENER_CONF_SECTION = 'listener'
 SERVICE_GROUP_CONF_SECTION = 'service_group'
 SERVER_CONF_SECTION = 'server'
+VIRTUAL_SERVER_CONFG_SECTION = 'slb'
```
Comment (afrin-chakure-a10) [inline a10_octavia/tests/common/a10constants.py:56]: Typo : `VIRTUAL_SERVER_CONFG_SECTION` -- > `VIRTUAL_SERVER_CONFIG_SECTION`

File: a10_octavia/tests/unit/controller/worker/tasks/test_virtual_server_tasks.py [hunk 9]
```diff
@@ -20,26 +21,84 @@
...
+        virtual_server_task = task.CreateVirtualServerTask()
+        virtual_server_task.axapi_client = self.client_mock
+        virtual_server_task.execute(LOADBALANCER, vthunder, flavor_data=flavor)
+        args, kwargs = self.client_mock.slb.virtual_server.create.call_args
+        self.assertEqual(kwargs['virtual_server']['arp_disable'], 1)
+
+    @mock.patch('a10_octavia.controller.worker.tasks.utils.parse_name_expressions',
+                mock.MagicMock())
+    def test_CreateVirtualServerTask_execute_flavor_override_config(self):
+        self.conf.config(group=a10constants.VIRTUAL_SERVER_CONFG_SECTION,
```
Comment (afrin-chakure-a10) [inline a10_octavia/tests/unit/controller/worker/tasks/test_virtual_server_tasks.py:72]: Typo : `VIRTUAL_SERVER_CONFG_SECTION` -- > `VIRTUAL_SERVER_CONFIG_SECTION`

File: tox.ini [hunk 10]
```diff
@@ -26,12 +26,14 @@ usedevelop = True
 install_command = pip install -U {opts} {packages}
 deps = -r{toxinidir}/requirements.txt
        -r{toxinidir}/test-requirements.txt
+       sqlalchemy_utils>0.33.0,<=0.34.0
```
Comment (afrin-chakure-a10) [inline tox.ini:29]: You can remove this file since its not related to FLAVORs

File: setup.cfg [hunk 11]
```diff
@@ -4,6 +4,9 @@ summary = A10 Octavia Provider Driver
 author = A10 Networks
 author-email = opensource@a10networks.com
 home-page = https://github.com/a10networks/a10-octavia
+long_description_content_type = text/markdown
```
Comment (afrin-chakure-a10) [inline setup.cfg:7]: Is this change made by you?

File: a10_octavia/controller/worker/tasks/utils.py [hunk 12]
```diff
@@ -73,3 +74,28 @@ def shared_template_modifier(template_type, template_name, device_templates):
...
+    flavor_data = {}
+    if name_expressions:
+        for expression in name_expressions:
+            if 'regex' in expression:
+                if re.search(expression['regex'], name):
+                    flavor_data.update(expression['json'])
+    return flavor_data
+
+
+def dash_to_underscore(my_dict):
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/tasks/utils.py:89]: Do we need unit test for this?

**Review by afrin-chakure-a10 — COMMENTED** [review]
All changes looks fine. Noticed that while basing this towards master,  some unrelated files are introduced.  You can remove them.

File: setup.cfg (hunk 11 above)
Comment (ytsai-a10) [reply setup.cfg:7]: This is change by Hunter. Maybe we can wait him back to check this.

File: tox.ini (hunk 10 above)
Comment (ytsai-a10) [reply tox.ini:29]: This is add by Hunter and I see some error after remove it when running "sudo tox -e py27".

File: ChangeLog (hunk 2 above)
Comment (ytsai-a10) [reply ChangeLog:4]: This is add by Hunter, maybe we can check this with him later.

File: AUTHORS (hunk 3 above)
Comment (ytsai-a10) [reply AUTHORS:5]: This is add by Hunter, maybe we can check this with him later.

File: ChangeLog (hunk 2 above)
Comment (afrin-chakure-a10) [reply ChangeLog:4]: ok

File: AUTHORS (hunk 3 above)
Comment (hthompson-a10) [reply AUTHORS:5]: Yeah let's remove it. These are auto added or something due to pbr tool.

File: ChangeLog (hunk 2 above)
Comment (hthompson-a10) [reply ChangeLog:4]: Yeah let's remove it. These are auto added or something due to pbr tool

**Review by afrin-chakure-a10 — DISMISSED** [review]

File: AUTHORS (hunk 3 above)
Comment (ytsai-a10) [reply AUTHORS]: removed

**Review by hthompson-a10 — DISMISSED** [review]

File: a10_octavia/common/config_options.py [hunk 13]
```diff
@@ -48,6 +48,9 @@
     cfg.BoolOpt('use_shared_for_template_lookup',
                 default=False,
                 help=_('Use shared for template')),
+    cfg.BoolOpt('use_regex_template_lookup',
```
Comment (ytsai-a10) [inline a10_octavia/common/config_options.py:51]: don't need this

**Review by rahulkatre-a10 — APPROVED** [review]

**Review by hthompson-a10 — APPROVED** [review]

**Pull request merged**